Build container and release it to github, merge files to pyproject.toml and handle basic setup/migrations automatically in container entrypoint-script

ilkka-ollakka opened the pull request:
<!--
Thanks for contributing! This template has some checkboxes that help keep track of what changes go into a release.

To check (tick) a list item, replace the space between square brackets with an x, like this:

- [x] I have checked the box

You can find more information and tips for BookWyrm contributors at https://docs.joinbookwyrm.com/contributing.html
-->
## Description
<!--
Describe what your pull request does here
-->
This PR contains changes to build bookwyrm container and release it to be used without closing the source repository. Currently it builds arm64 and amd64 platforms.

Container builds are done when tag is pushed to use tag as version, for example `bookwyrm:0.8.0`. Main repo push build container with name `bookwyrm:main`.

It also combines requirements.txt, pylint/mypy/coverage configs to pyproject.toml config.

Container entrypoint script runs migrations,initdb and permissions checks before changing to `bookwyrm`-user to start provided command. So Celery/bookwyrm services are running non-root user. I used the https://github.com/bookwyrm-social/bookwyrm/pull/2467 as initial entrypoint script and modified it littlebit.

`bw-dev` command now has command `dev` to start local code in container instead of build container. It should allow local development easily. `dev`-config also sets `DEBUG=1` environment variable, making heavy assumption that you want to run it with debug-mode.

> One flow is not yet handled properly, that is that celery doesn't wait migrations after initial migrations, so new migrations are done, but celery/others doesn't know to wait them to finish before starting.

I modified Dockerfile and entrypoint for this, to take build-timestamp as marker what to look for in file when waiting migration checks, still might not be fully working, but should handle new builds/migrations to be waited correctly.

Changes include using dev-tools container changes, build it on top of bookwyrm container, install all dev-dependencies there instead of main container and run all pytest/mypy etc using that container.

<!--
For pull requests that relate or close an issue, please include them
below.  We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "closes #1234" would connect the current pull
request to issue 1234.  And when we merge the pull request, Github will
automatically close the issue.
-->

- Related Issue #2095 
- Closes #3551 

## What type of Pull Request is this?
<!-- Check all that apply -->

- [ ] Bug Fix
- [X] Enhancement
- [x] Plumbing / Internals / Dependencies
- [X] Refactor

## Does this PR change settings or dependencies, or break something?
<!-- Check all that apply -->

- [ ] This PR changes or adds default settings, configuration, or .env values
- [ ] This PR changes or adds dependencies
- [X] This PR introduces other breaking changes

### Details of breaking or configuration changes (if any of above checked)
We don't anymore mount local directory to container, but build container. We also run migrations and database-initdb autotically on each start.

We run pytest/mypy/etc in derived container that mounts source-code in.

Dev flow now uses `bw-dev dev` instead of `bw-dev up` to start local code.

`requirements.txt`, `.pylintrc`, `.coveragerc` and `mypy.ini` are all merged to `pyproject.toml`

## Documentation
<!--
Documentation for users, admins, and developers is an important way to keep the BookWyrm community welcoming and make Bookwyrm easy to use.
Our documentation is maintained in a separate repository at https://github.com/bookwyrm-social/documentation
-->

<!-- Check all that apply -->

- [X] New or amended documentation will be required if this PR is merged
- [ ] I have created a matching pull request in the Documentation repository
- [ ] I intend to create a matching pull request in the Documentation repository after this PR is merged

<!-- Amazing! Thanks for filling that out. Your PR will need to have passing tests and happy linters before we can merge
You will need to check your code with `black`, `pylint`, and `mypy`, or `./bw-dev formatters`
-->

### Tests
<!-- Check one -->

- [ ] My changes do not need new tests
- [ ] All tests I have added are passing
- [ ] I have written tests but need help to make them pass
- [ ] I have not written tests and need help to write them

I think this needs some manual testing, as it is big change and it is probable that I have missed some flow/use-case that is broken.

osresearch committed:
docker: move initial setup into docker startup to simplify deployment

ilkka-ollakka committed:
Rework Dockerfile to create smaller container and run as non-root by default

This changes the created image size from ~1.3G -> ~0.5G  and runs code as non-root uid in container.

Also creates dev-tool container that container pytest/pylint/tooling and it is used to run those
instead of main container.

In dev-tools use npm-18 container to copy runtime instead of repository. Making it more visible what
version is used.

arm64 side doesn't have same binary support in wheel files as amd64 has, so better keep build image so we can build needed dependencies if needed

move entrypoint-script outside bookwyrm directory and modify flow
- Check parameters given for entrypoint and gunicorn runner is the web-instance, so we don't run migration checks with run-web
  calls
- Always run migrations check in web-node
- Change user to bookwyrm before starting to execute command

ilkka-ollakka committed:
bw-dev: run tooling as root-uid in bookwyrm-dev-tools -container

We keep the tooling in separate container which is build on top of bookwyrm container.

ilkka-ollakka committed:
Add github action to build and release containers

ilkka-ollakka committed:
merge requirements/mypy/pylint/coverage to pyproject.toml

ilkka-ollakka committed:
fix dev-tools mounts

otherwise it would shadow installed node_modules location

ilkka-ollakka committed:
Dockerfile: add buildtime hash of migrations as info to wait for migrations

Instead of checking file-presence, check that file contains build-time define hash of migration
files

This allows newer builds/migrations to be waited until those are checked properly

ilkka-ollakka commented:
rebased the PR on top of current main

ilkka-ollakka commented on Dockerfile:
```diff
@@ -1,12 +1,43 @@
+COPY locale /app/locale
+COPY bookwyrm /app/bookwyrm
+COPY entrypoint.sh /entrypoint.sh
+RUN date +%s > /build_timestamp
```
Instead of plain timestamp, we could run for example hash of migration-files hashes to have identifier. Biggest difference would be that it wouldn't recheck the migrations if there is no changes, so in release flow not huge difference.

ilkka-ollakka replied:
I changed the entrypoint/Dockerfile for hash-flow instead of build-timestamp.

ilkka-ollakka committed:
bw-dev: add 'dev' command to run local code instead of build container

docker-compose.dev.yml mounts most of the files on top of container, so you can develop entrypoint.sh as well as bookwyrm code/locales etc with dev command.

ilkka-ollakka committed:
docker-compose: change bookwyrm image to point to ghcr.io

ilkka-ollakka committed:
docker-compose: move dev-tools container to dev-compose-file

marcusyoung commented:
@ilkka-ollakka Can I check that consideration has been given to any potential impact of these changes on dockerless installs. For example, I noticed that requirements.txt has been removed, But we need to run `./venv/bin/pip3 install -r requirements.txt` on a dockerless install during upgrade. There may be other things, but that's the one that stood out.

ilkka-ollakka commented:
I'll split separate PR's out from this, separatable changes so those are easier to review/merge separately after current release is in order.

ilkka-ollakka committed:
Dockerfile: remove expose port

we only expose that port in web role, not in celery etc containers

ilkka-ollakka committed:
docker-compose: define healthcheck to web-container and simplify entrypoint

Celery and flower containers wait until web container is healthy, so it has handled migrations and database init

This allows to simplify entrypoint-script so it doesn't need to poll in other containers if migrations are done

ilkka-ollakka committed:
entrypoint: add missed compile_themes step

SensibleSalmon commented on pyproject.toml:
```diff
@@ -1,2 +1,151 @@
+    "ALLOWED_HOSTS = your.domain.here",
+    "BOOKWYRM_DATABASE_BACKEND = postgres",
+    "MEDIA_ROOT = images/",
+    "CELERY_BROKER_URL = memory://",
```
curiosity: is this meant to be a default? in the docker-compose case, would `memory://` ever work when the broker is a separate container?

ilkka-ollakka replied:
this is oblyfor pytest run variables, so not default otherwise. It is in tool pytest.ini_options block

ilkka-ollakka commented on docker-compose.yml:
```diff
@@ -45,20 +46,23 @@ services:
       - main
   web:
+    image: ghcr.io/bookwyrm/bookwyrm:main
     build: .
```
allowing development and local build. Also as all have same image it only builds it once, atleast podman compose builds currently all images separately.

hughrun commented:
@ilkka-ollakka is the idea that we won't actually need this PR once we merge #3803 #3809 plus maybe some small cleanup after that?

ilkka-ollakka committed:
Merge branch 'main' into feat/container_build_and_release

ilkka-ollakka commented:
closing this PR as relevant changes are merged from PR splitted out from this.
